pahfit/fitter.py

```diff
@@ -0,0 +1,203 @@
+from abc import ABC, abstractmethod
+
+
+class Fitter(ABC):
+    """Abstract base class for interal Fitter API.
+
+    All shared methods should have the same arguments, enforced by this
+    abstract class. Any API-specific options preferably go into the
+    constructor of the subclass, although some general-purpose
+    dictionaries could also be used if absolutely necessary.
+
+    The main functionalities of a Fitter subclass:
+    1. Convert the numbers that are in the Features table to a fittable
+       model configuration for a certain framework. The details of the
+       fitting framework are hidden behind the respective subclass.
+    2. Fit the model to the spectrum without any additional assumptions.
+       The Fitter will fit the given data using the given model without
+       thinking about redshift, units, instrumental effects.)
+    3. Retrieve the fitted quantities, which are the values that were
+       passed during step 1. When fit result uncertainties are
+       implemented, they will also need to be retrieved through this
+       API.
+    4. Access to the evaluation of the underlying model (again with no
+       assumptions like in step 2.).
+
+    For the model setup, multiple functions are used, so a few notes are
+    provided here. There is one function per type of component supported
+    by PAHFIT, and the arguments of these functions will ask for
+    different "standard" PAHFIT numbers, i.e. those from the Features
+    table. These functions have the same signature between all Fitter
+    implementations, so that the Model class can use a single
+    implementation to set up the Fitter. The Model has access to the
+    Features table and the instrument model, and needs to set up Fitter
+    with the correct initial values, bounds, and "fixed" flags (e.g.
+    setting a fixed FWHM based on the instrument for the lines). After
+    all the components have been added, the finalize_model() function
+    can be called to finish setting up the internal astropy model. After
+    this has finished, fit() can be called to apply the model and the
+    astropy fitter to the data.
```

Comment:
Pretty confused about this description, and why a "finalize" is needed.  It also references AstroPy.

Reply:
So is `Model` just "spoon-feeding" the rows of the `Features` table to a `Fitter` object?

Reply:
Yes. While it might not look like the most efficient way, I felt like this way would lead to the least assumptions. Basically, I tried avoiding doing anything "smart", so that the main focus could lie on the providing a uniform interface to the setup, fitting, and result retrieval mechanics (the things that will always be different between fitting back-ends)

I admit that it does reflect the feel of the underlying astropy: you create individual components and sum them together. But to set up each components, the right numbers need to be derived from the contents of the features table.

Reply:
I think this could be workable, as long as we can re-use a `Fitter` when requested.